### DISCUSSION TITLE
Support for more complex bitfields

### DISCUSSION THREAD

**Discussion by qtc-de** [vlang/v#18586] [Ideas]
Hi there :wave: 

I'm dealing with a C structure that looks like this:

```c
struct some_struct {
  unsigned int field_1 : 1;
  unsigned int field_2 : 1;
  unsigned int field_3 : 1;
  unsigned int field_4 : 1;
  unsigned int field_5 : 12;
  unsigned int field_6 : 8;
  unsigned int field_7 : 8;
  ... other fields ...
}
```

Using bitfields works great for structures where all fields take a single bit, but I have not found an example how to deal with a structure like the one above. Is there any way to represent this in V?

Best
Tobias

**Comment by mateli** [comment]
In embedded programming accessing memory mapped registers are very important and that usually requires a packed struct which different compilers has different syntax for. The usual solution is to define a more portable PACKED attribute and using it like this:
PACKED typedef struct {
  uint16_t field_1 : 1;
  uint16_t field_2 : 1;
  uint16_t field_3 : 1;
  uint16_t field_4 : 1;
  uint16_t field_5 : 12;
  uint16_t field_6 : 8;
  uint16_t field_7 : 8;
  ... other fields ...
} some_struct;

An alternative syntax would be something like this.
packed struct some_struct : uint16 {
  uint1 field_1 : 1;
  uint1 field_2 : 1;
  uint1 field_3 : 1;
  uint1 field_4 : 1;
  uint12 field_5 : 12;
  uint8 field_6 : 8;
  uint8 field_7 : 8;
  ... other fields ...
};
The uint16 here would mean that the struct is padded by the bits needed to fill a multiple of 16 bits. That translates nicely to the C standard above. A cool option would be to for example map it to a uint15 which doesn't cleanly map to C code but would provide a feature that is missing in C. Also a 16bit struct should be assignable to an uint16 and a larger struct to an uint16 array without having to use unions or casting tricks. 

This is also important for other use cases such as dealing with binary files where you can memory map a file to memory and then just take a known pointer within that an cast it to a struct pointer and work with the the fields that is expected to be in a file. Network communication heavily depends on packed structs.

Furthermore there would be nice to have some that C does not have. Such as non-pointer addressing of fields in a struct. 
fields some_struct_fields = {some_struct.field_3, somestruct_struct.field_2, somestruct.field_5};
void change_field(int num, int value) {
    some_struct_fields[num] = value;
}
Something like that would save us from writing lots of select/case code. 

The default C behavior is to truncate integers that is to long without any warning or error. Unless the integer is set to a constant value during initialization. This behavior is different from for example setting an uint16 to an uint32 value which implies casting and causes an error if the uint32 it to large which makes it less safe but faster.
